README.md
# KostalPyko
This package is for working with a Piko (NOT Plenticore) Inverter from [Kostal](http://www.kostal-solar-electric.com/)

## Where has it been tested?
KostalPyko has been tested with a Kostal Piko 5.5. It should work with other Kostal inverters as well since they provide the same interface.

## Requirements
 * Requests
 * Python
 * lxml
 * httplib2
 * httpretty (for testing)
 
## Installing
```bash
$ pip install kostalpyko
```

## Usage
    from kostalpyko.kostalpyko import Piko
    
    #create a new piko instance
    p = Piko('host', 'username', 'password')
    
    #get current power 
    print(p.get_current_power())
    
    #get voltage from string 1
    print(p.get_string1_voltage())

## Thanks to
https://github.com/Tafkas since this packet is based on https://github.com/Tafkas/KostalPikoPy


CHANGES.txt
v0.1, 2014-07-31 -- Initial release.
v0.2, 2020-11-16 -- Fixed some bugs.
v0.3, 2021-11-29 -- Fixed some bugs with indices.
v0.4, 2022-05-24 -- Fixed some bugs with indices and added utility to safely get values.
v0.5, 2022-08-22 -- Added compatibility to inverters with 3 phases 2 strings.
v0.5.1, 2022-08-22 -- Some fixes.
v0.6, 2022-08-25 -- Added compatibility to inverters with 1 phases 2 strings.

tests/test_kostalpiko.py
import httpretty
import unittest
from kostalpyko.kostalpyko import Piko


class Test(unittest.TestCase):
    def setUp(self):
        httpretty.enable()  # enable HTTPretty so that it will monkey patch the socket module
        httpretty.register_uri(httpretty.GET, "http://example.com", body=open("fixtures/index.html").read())

    def tearDown(self):
        httpretty.disable()  # disable afterwards, so that you will have no problems in code that uses that socket module
        httpretty.reset()  # reset HTTPretty state (clean up registered urls and request history)

    def test_get_raw_content(self):
        p = Piko(host='http://example.com')
        self.assertEqual(p._get_raw_content(),
                         ['112', '9290', '19.83', '384', '230', '0.20', '0', '278', '232', '0.21', '112', '0', '230',
                          '0.00', '0'])

    def test_get_current_power(self):
        p = Piko(host='http://example.com')
        self.assertEqual(p.get_current_power(), 112)

    def test_get_total_energy(self):
        p = Piko(host='http://example.com')
        self.assertEqual(p.get_total_energy(), 9290)

    def test_get_daily_energy(self):
        p = Piko(host='http://example.com')
        self.assertEqual(p.get_daily_energy(), 19.83)

    def test_get_string1_voltage(self):
        p = Piko(host='http://example.com')
        self.assertEqual(p.get_string1_voltage(), 384)

    def test_get_string2_voltage(self):
        p = Piko(host='http://example.com')
        self.assertEqual(p.get_string2_voltage(), 278)

    def test_get_string3_voltage(self):
        p = Piko(host='http://example.com')
        self.assertEqual(p.get_string3_voltage(), 0)

    def test_get_string1_current(self):
        p = Piko(host='http://example.com')
        self.assertEqual(p.get_string1_current(), 0.2)

    def test_get_string2_current(self):
        p = Piko(host='http://example.com')
        self.assertEqual(p.get_string2_current(), 0.21)

    def test_get_string3_current(self):
        p = Piko(host='http://example.com')
        self.assertEqual(p.get_string3_current(), 0.0)

    def test_get_l1_voltage(self):
        p = Piko(host='http://example.com')
        self.assertEqual(p.get_l1_voltage(), 230)

    def test_get_l2_voltage(self):
        p = Piko(host='http://example.com')
        self.assertEqual(p.get_l2_voltage(), 232)

    def test_get_l3_voltage(self):
        p = Piko(host='http://example.com')
        self.assertEqual(p.get_l3_voltage(), 230)

    def test_get_l1_power(self):
        p = Piko(host='http://example.com')
        self.assertEqual(p.get_l1_power(), 0)

    def test_get_l2_power(self):
        p = Piko(host='http://example.com')
        self.assertEqual(p.get_l2_power(), 112)

    def test_get_l3_power(self):
        p = Piko(host='http://example.com')
        self.assertEqual(p.get_l3_power(), 0)


tests/test_kostalpiko_2_1.py
import httpretty
import unittest
from kostalpiko.kostalpiko import Piko
import os

class Test551(unittest.TestCase):
    
    def setUp(self):
        httpretty.enable(verbose=True)  # enable HTTPretty so that it will monkey patch the socket module
        httpretty.register_uri(httpretty.GET, "http://example.com/index.fhtml", body=open("./tests/fixtures/index2-1.html").read())
        self.piko = Piko(host='http://example.com')
        self.piko.update_data()
        # print("Test 2-1")

    def tearDown(self):
        httpretty.disable()  # disable afterwards, so that you will have no problems in code that uses that socket module
        httpretty.reset()  # reset HTTPretty state (clean up registered urls and request history)

    def test_get_raw_content(self):
        self.assertEqual(self.piko.data._raw_data,
                         ['2336', '33523', '11.71', '338', '236', '3.88', '2336', '336', '3.97', 'supply MPP'])

    def test_get_current_power(self):
        self.assertEqual(self.piko.data.get_current_power(), 2336)

    def test_get_total_energy(self):
        self.assertEqual(self.piko.data.get_total_energy(), 33523)

    def test_get_daily_energy(self):
        self.assertEqual(self.piko.data.get_daily_energy(), 11.71)

    def test_get_string1_voltage(self):
        self.assertEqual(self.piko.data.get_string1_voltage(), 338)

    def test_get_string2_voltage(self):
        self.assertEqual(self.piko.data.get_string2_voltage(), 336)

    def test_get_string1_current(self):
        self.assertEqual(self.piko.data.get_string1_current(), 3.88)

    def test_get_string2_current(self):
        self.assertEqual(self.piko.data.get_string2_current(), 3.97)

    def test_get_l1_voltage(self):
        self.assertEqual(self.piko.data.get_l1_voltage(), 236)

    def test_get_l1_power(self):
        self.assertEqual(self.piko.data.get_l1_power(), 2336)

    def status(self):
        self.assertEqual(self.piko.data.get_status(), 'supply MPP')


tests/test_kostalpiko_5_5_1.py
import httpretty
import unittest
from kostalpiko.kostalpiko import Piko
import os

class Test551(unittest.TestCase):
    
    def setUp(self):
        httpretty.enable(verbose=True)  # enable HTTPretty so that it will monkey patch the socket module
        httpretty.register_uri(httpretty.GET, "http://example.com/index.fhtml", body=open("./tests/fixtures/index55-1.html").read())
        self.piko = Piko(host='http://example.com')
        self.piko.update_data()
        # print("Test 5-5-1")

    def tearDown(self):
        httpretty.disable()  # disable afterwards, so that you will have no problems in code that uses that socket module
        httpretty.reset()  # reset HTTPretty state (clean up registered urls and request history)

    def test_get_raw_content(self):
        self.assertEqual(self.piko.data._raw_data,
                         ['112', '9290', '19.83', '384', '230', '0.20', '0', '278', '232', '0.21', '112', '0', '230',
                          '0.00', '0', 'Einspeisen MPP'])

    def test_get_current_power(self):
        self.assertEqual(self.piko.data.get_current_power(), 112)

    def test_get_total_energy(self):
        self.assertEqual(self.piko.data.get_total_energy(), 9290)

    def test_get_daily_energy(self):
        self.assertEqual(self.piko.data.get_daily_energy(), 19.83)

    def test_get_string1_voltage(self):
        self.assertEqual(self.piko.data.get_string1_voltage(), 384)

    def test_get_string2_voltage(self):
        self.assertEqual(self.piko.data.get_string2_voltage(), 278)

    def test_get_string3_voltage(self):
        self.assertEqual(self.piko.data.get_string3_voltage(), 0)

    def test_get_string1_current(self):
        self.assertEqual(self.piko.data.get_string1_current(), 0.2)

    def test_get_string2_current(self):
        self.assertEqual(self.piko.data.get_string2_current(), 0.21)

    def test_get_string3_current(self):
        self.assertEqual(self.piko.data.get_string3_current(), 0.0)

    def test_get_l1_voltage(self):
        self.assertEqual(self.piko.data.get_l1_voltage(), 230)

    def test_get_l2_voltage(self):
        self.assertEqual(self.piko.data.get_l2_voltage(), 232)

    def test_get_l3_voltage(self):
        self.assertEqual(self.piko.data.get_l3_voltage(), 230)

    def test_get_l1_power(self):
        self.assertEqual(self.piko.data.get_l1_power(), 0)

    def test_get_l2_power(self):
        self.assertEqual(self.piko.data.get_l2_power(), 112)

    def test_get_l3_power(self):
        self.assertEqual(self.piko.data.get_l3_power(), 0)
    
    def status(self):
        self.assertEqual(self.piko.data.get_status(), 'Einspeisen MPP')


pikopy/piko.py
#!/usr/bin/env python

# The MIT License (MIT)
#
# Copyright (c) 2014 Christian Stade-Schuldt
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

"""Library to work with a Piko inverter from Kostal."""

from sys import version_info
from lxml import html

# HTTP libraries depends upon Python 2 or 3
if version_info.major == 3:
    import requests
else:
    from urllib import urlencode
    import urllib2


class Piko():
    def __init__(self, host=None, username='pvserver', password='pvwr'):
        self.host = host
        self.username = username
        self.password = password

    def get_solar_generator_power(self):
        """returns the current power of the solar generator in W"""
        if self._get_content_of_own_consumption() != []:
            return self._get_content_of_own_consumption()[5]
        else:
            return "No BA sensor installed"

    def get_consumption_phase_1(self):
        """returns the current consumption of phase 1 in W"""
        if self._get_content_of_own_consumption() != []:
            return self._get_content_of_own_consumption()[8]
        else:
            return "No BA sensor installed"

    def get_consumption_phase_2(self):
        """returns the current consumption of phase 2 in W"""
        if self._get_content_of_own_consumption() != []:
            return self._get_content_of_own_consumption()[9]
        else:
            return "No BA sensor installed"

    def get_consumption_phase_3(self):
        """returns the current consumption of phase 3 in W"""
        if self._get_content_of_own_consumption() != []:
            return self._get_content_of_own_consumption()[10]
        else:
            return "No BA sensor installed"

    def _get_content_of_own_consumption(self):
        """returns all values as a list"""
        if version_info.major == 3:
            url = self.host + '/BA.fhtml'
            login = self.username
            pwd = self.password
            try:
                r = requests.get(url, auth=(login, pwd), timeout=15)
                if r.status_code == 200:
                    response = html.fromstring(r.content)
                    data = []
                    for v in response.xpath("//b"):
                        raw = v.text.strip()
                        raw = raw[:-1]  # remove unit
                        try:
                            value = float(raw)
                        except:
                            value = 0
                        data.append(value)
                    return data
                else:
                    raise ConnectionError
            except requests.exceptions.ConnectionError as errc:
                return None
            except requests.exceptions.Timeout as errt:
                return None

        else:
            url = self.host + "/BA.fhtml"
            password_mgr = urllib2.HTTPPasswordMgrWithDefaultRealm()
            password_mgr.add_password(
                None, self.host, self.username, self.password)
            handler = urllib2.HTTPBasicAuthHandler(password_mgr)
            opener = urllib2.build_opener(handler)
            opener.open(url)

            urllib2.install_opener(opener)
            response = urllib2.urlopen(self.host)
            root = html.fromstring(response.read().strip())

            data = []
            for v in root.xpath("//b"):
                raw = v.text.strip()
                raw = raw[:-1]  # remove unit
                try:
                    value = float(raw)
                except:
                    value = 0
                data.append(value)

            return data

    def get_logdaten_dat(self):
        pass

    def get_current_power(self):
        """returns the current power in W"""
        if self._get_raw_content() is not None:
            return int(self._get_raw_content()[0])

    def get_total_energy(self):
        """returns the total energy in kWh"""
        if self._get_raw_content() is not None:
            return int(self._get_raw_content()[1])

    def get_daily_energy(self):
        """returns the daily energy in kWh"""
        if self._get_raw_content() is not None:
            return float(self._get_raw_content()[2])

    def get_string1_voltage(self):
        """returns the voltage from string 1 in V"""
        if self._get_raw_content() is not None:
            return int(self._get_raw_content()[3])

    def get_string1_current(self):
        """returns the current from string 1 in A"""
        if self._get_raw_content() is not None:
            return float(self._get_raw_content()[5])

    def get_string2_voltage(self):
        """returns the voltage from string 2 in V"""
        if self._get_raw_content() is not None:
            return int(self._get_raw_content()[7])

    def get_string2_current(self):
        """returns the current from string 2 in A"""
        if self._get_raw_content() is not None:
            return float(self._get_raw_content()[9])

    def get_string3_voltage(self):
        """returns the voltage from string 3 in V"""
        raw_content = self._get_raw_content()
        if len(raw_content) < 14:
            # String 3 not installed
            return None
        else:
            return int(raw_content[11])

    def get_string3_current(self):
        """returns the current from string 3 in A"""
        raw_content = self._get_raw_content()
        if len(raw_content) < 14:
            # String 3 not installed
            return None
        else:
            return float(raw_content[13])

    def get_l1_voltage(self):
        """returns the voltage from line 1 in V"""
        if self._get_raw_content() is not None:
            return int(self._get_raw_content()[4])

    def get_l1_power(self):
        """returns the power from line 1 in W"""
        if self._get_raw_content() is not None:
            return int(self._get_raw_content()[6])

    def get_l2_voltage(self):
        """returns the voltage from line 2 in V"""
        if self._get_raw_content() is not None:
            return int(self._get_raw_content()[8])

    def get_l2_power(self):
        """returns the power from line 1 in W"""
        if self._get_raw_content() is not None:
            return int(self._get_raw_content()[10])

    def get_l3_voltage(self):
        """returns the voltage from line 3 in V"""
        return int(self._get_raw_content()[12])

    def get_l3_power(self):
        """returns the power from line 3 in W"""
        return int(self._get_raw_content()[14])

    def _get_raw_content(self):
        """returns all values as a list"""
        if version_info.major == 3:
            url = self.host + '/index.fhtml'
            login = self.username
            pwd = self.password
            try:
                r = requests.get(url, auth=(login, pwd), timeout=15)
                if r.status_code == 200:
                    response = html.fromstring(r.content)
                    data = []
                    for v in response.xpath("//td[@bgcolor='#FFFFFF']"):
                        raw = v.text.strip()
                        if ('x x x' in raw):
                            raw = 0
                        data.append(raw)
                    return data
                else:
                    raise ConnectionError
            except requests.exceptions.ConnectionError as errc:
                return None
            except requests.exceptions.Timeout as errt:
                return None
        else:
            password_mgr = urllib2.HTTPPasswordMgrWithDefaultRealm()
            password_mgr.add_password(
                None, self.host, self.username, self.password)
            handler = urllib2.HTTPBasicAuthHandler(password_mgr)
            opener = urllib2.build_opener(handler)
            opener.open(self.host)

            urllib2.install_opener(opener)
            response = urllib2.urlopen(self.host)
            root = html.fromstring(response.read().strip())
            data = [v.text.strip() for v in root.xpath(
                "//td[@bgcolor='#FFFFFF']")]
            return data


kostalpiko/const.py
BASE_INDICES = {
    "current_power": 0,
    "total_energy": 1,
    "daily_energy": 2,
    "string1_voltage": 3,
    "l1_voltage": 4,
    "string1_current": 5,
    "l1_power": 6,
}

SINGLE_STRING_INDICES = {
    **BASE_INDICES,
    "status": 7
}

DOUBLE_STRING_SINGLE_PHASE_INDICES = {
    **BASE_INDICES,
    "string2_voltage": 7,
    "string2_current": 8,
    "status": 9
}

DOUBLE_STRING_TWO_PHASES_INDICES = {
    **BASE_INDICES,
    "string2_voltage": 7,
    "l2_voltage": 8,
    "string2_current": 9,
    "l2_power": 10,
    "status": 11 
}

DOUBLE_STRING_THREE_PHASES_INDICES = {
    **BASE_INDICES,
    "string2_voltage": 7,
    "l2_voltage": 8,
    "string2_current": 9,
    "l2_power": 10,
    "l3_voltage": 11,
    "l3_power": 12,
    "status": 13 
}

TRIPLE_STRING_INDICES = {
    **BASE_INDICES,
    "string2_voltage": 7,
    "l2_voltage": 8,
    "string2_current": 9,
    "l2_power": 10,
    "string3_voltage": 11,
    "l3_voltage": 12,
    "string3_current": 13,
    "l3_power": 14,
    "status": 15 
}


kostalpiko/kostalpiko.py
#!/usr/bin/env python


"""Library to work with a Piko inverter from Kostal."""
import logging
import requests
from lxml import html
from .utils import safe_list_get
from .const import *

LOG = logging.getLogger(__name__)

class Piko:
    def __init__(self, host=None, username="pvserver", password="pvwr"):
        self.host = host
        self.username = username
        self.password = password
        self.data = None
        self.ba_data = None

    def update(self):
        """updates all data"""
        self.update_data()
        self.update_ba_data()

    def update_data(self):
        """updates data"""
        try:
            data = self._get_raw_content()
            if data is not None:
                self.data = PikoData(data)
        except Exception as err:
            LOG.debug(err)
            pass

    def update_ba_data(self):
        """updates ba data"""
        try:
            data = self._get_content_of_own_consumption()
            if data is not None:
                self.ba_data = PikoBAData(data)
        except Exception as err:
            LOG.debug(err)
            pass

    def _get_raw_content(self):
        """returns all values as a list"""
        url = self.host + "/index.fhtml"
        login = self.username
        pwd = self.password
        try:
            r = requests.get(url, auth=(login, pwd), timeout=15)
            if r.status_code == 200:
                response = html.fromstring(r.content)
                data = []
                for v in response.xpath("//td[@bgcolor='#FFFFFF']"):
                    raw = v.text.strip()
                    if "x x x" in raw:
                        raw = 0
                    data.append(raw)
                status = response.xpath("/html/body/form/font/table[2]/tr[8]/td[3]")[
                    0
                ].text.strip()
                data.append(status)
                LOG.debug(data)
                return data
            else:
                raise ConnectionError
        except requests.exceptions.ConnectionError as errc:
            LOG.error(errc)
            return None
        except requests.exceptions.Timeout as errt:
            LOG.error(errt)
            return

    def _get_content_of_own_consumption(self):
        """returns all values as a list"""
        url = self.host + "/BA.fhtml"
        login = self.username
        pwd = self.password
        try:
            r = requests.get(url, auth=(login, pwd), timeout=15)
            if r.status_code == 200:
                response = html.fromstring(r.content)
                data = []
                for v in response.xpath("//b"):
                    raw = v.text.strip()
                    raw = raw[:-1]  # remove unit
                    try:
                        value = float(raw)
                    except:
                        value = 0
                    data.append(value)
                LOG.debug(data)
                return data
            else:
                raise ConnectionError
        except requests.exceptions.ConnectionError as errc:
            LOG.debug(errc)
            return None
        except requests.exceptions.Timeout as errt:
            LOG.debug(errt)
            return None

    def _get_info(self):
        """returns the info about the inverter"""
        url = self.host + "/Solar2.fhtml"
        login = self.username
        pwd = self.password
        try:
            r = requests.get(url, auth=(login, pwd), timeout=15)
            if r.status_code == 200:
                response = html.fromstring(r.content)
                data = []
                serial = response.xpath("/html/body/form/font/table/tr[2]/td[3]")[
                    0
                ].text.strip()
                data.append(serial)
                model = response.xpath("/html/body/form/table/tr[2]/td[2]/font[1]")[
                    0
                ].text.strip()
                data.append(model)
                LOG.debug(data)
                return data
            else:
                raise ConnectionError
        except requests.exceptions.ConnectionError as errc:
            LOG.debug(errc)
            return None
        except requests.exceptions.Timeout as errt:
            LOG.debug(errt)
            return None

    def get_logdaten_dat(self):
        pass


class PikoData(object):
    """
    PIKO Data
    """

    def __init__(self, raw_data):
        self._raw_data = raw_data
        n_values = len(raw_data)
        if n_values == 8:
            self.indices = SINGLE_STRING_INDICES
        elif n_values == 10:
            self.indices = DOUBLE_STRING_SINGLE_PHASE_INDICES
        elif n_values == 12:
            self.indices = DOUBLE_STRING_TWO_PHASES_INDICES
        elif n_values == 14:
            self.indices = DOUBLE_STRING_THREE_PHASES_INDICES
        elif n_values == 16:
            self.indices = TRIPLE_STRING_INDICES

    def safe_get_value(self, name):
        if self._raw_data is not None and name in self.indices and self.indices[name] is not None:
            return safe_list_get(self._raw_data, self.indices[name])

    def get_current_power(self):
        """returns the current power in W"""
        value = self.safe_get_value('current_power')
        if value is not None:
            return int(value)
        else:
            return None

    def get_total_energy(self):
        """returns the total energy in kWh"""
        value = self.safe_get_value('total_energy')
        if value is not None:
            return int(value)
        else:
            return None

    def get_daily_energy(self):
        """returns the daily energy in kWh"""
        value = self.safe_get_value('daily_energy')
        if value is not None:
            return float(value)
        else:
            return None

    def get_string1_voltage(self):
        """returns the voltage from string 1 in V"""
        value = self.safe_get_value('string1_voltage')
        if value is not None:
            return int(value)
        else:
            return None

    def get_string1_current(self):
        """returns the current from string 1 in A"""
        value = self.safe_get_value('string1_current')
        if value is not None:
            return float(value)
        else:
            return None

    def get_string2_voltage(self):
        """returns the voltage from string 2 in V"""
        value = self.safe_get_value('string2_voltage')
        if value is not None:
            return int(value)
        else:
            return None

    def get_string2_current(self):
        """returns the current from string 2 in A"""
        value = self.safe_get_value('string2_current')
        if value is not None:
            return float(value)
        else:
            return None

    def get_string3_voltage(self):
        """returns the voltage from string 3 in V"""
        value = self.safe_get_value('string3_voltage')
        if value is not None:
            return int(value)
        else:
            return None

    def get_string3_current(self):
        """returns the current from string 3 in A"""
        value = self.safe_get_value('string3_current')
        if value is not None:
            return float(value)
        else:
            return None

    def get_l1_voltage(self):
        """returns the voltage from line 1 in V"""
        value = self.safe_get_value('l1_voltage')
        if value is not None:
            return int(value)
        else:
            return None

    def get_l1_power(self):
        """returns the power from line 1 in W"""
        value = self.safe_get_value('l1_power')
        if value is not None:
            return int(value)
        else:
            return None

    def get_l2_voltage(self):
        """returns the voltage from line 2 in V"""
        value = self.safe_get_value('l2_voltage')
        if value is not None:
            return int(value)
        else:
            return None

    def get_l2_power(self):
        """returns the power from line 1 in W"""
        value = self.safe_get_value('l2_power')
        if value is not None:
            return int(value)
        else:
            return None

    def get_l3_voltage(self):
        """returns the voltage from line 3 in V"""
        value = self.safe_get_value('l3_voltage')
        if value is not None:
            return int(value)
        else:
            return None

    def get_l3_power(self):
        value = self.safe_get_value('l3_power')
        if value is not None:
            return int(value)
        else:
            return None

    def get_piko_status(self):
        """returns the power from line 3 in W"""
        value = self.safe_get_value('status')
        if value is not None:
            return value
        else:
            return None


class PikoBAData(object):
    """
    PIKO BA Data
    """

    def __init__(self, raw_data):
        self._raw_data = raw_data

    def get_solar_generator_power(self):
        """returns the current power of the solar generator in W"""
        if self._raw_data:
            return safe_list_get(self._raw_data, 5)
        else:
            return "No BA sensor installed"

    def get_consumption_phase_1(self):
        """returns the current consumption of phase 1 in W"""
        if self._raw_data:
            return safe_list_get(self._raw_data, 8)
        else:
            return "No BA sensor installed"

    def get_consumption_phase_2(self):
        """returns the current consumption of phase 2 in W"""
        if self._raw_data:
            return safe_list_get(self._raw_data, 9)
        else:
            return "No BA sensor installed"

    def get_consumption_phase_3(self):
        """returns the current consumption of phase 3 in W"""
        if self._raw_data:
            return safe_list_get(self._raw_data, 10)
        else:
            return "No BA sensor installed"
